zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/MultipleAddresses.java

```diff
@@ -0,0 +1,187 @@
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server.quorum;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.NoRouteToHostException;
+import java.net.UnknownHostException;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This class allows to store several quorum and electing addresses.
+ *
+ * See ZOOKEEPER-3188 for a discussion of this feature.
+ */
+public class MultipleAddresses {
+
+    private Set<InetSocketAddress> addresses;
+    private int timeout;
+
+    public MultipleAddresses() {
+        addresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        timeout = 100;
+    }
+
+    public MultipleAddresses(List<InetSocketAddress> addresses) {
+        this(addresses, 100);
+    }
+
+    public MultipleAddresses(InetSocketAddress address) {
+        this(address, 100);
+    }
+
+    public MultipleAddresses(List<InetSocketAddress> addresses, int timeout) {
+        this.addresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        this.addresses.addAll(addresses);
+        this.timeout = timeout;
+    }
+
+    public MultipleAddresses(InetSocketAddress address,  int timeout) {
+        addresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        addresses.add(address);
+        this.timeout = timeout;
+    }
+
+    public int getTimeout() {
+        return timeout;
+    }
+
+    public void setTimeout(int timeout) {
+        this.timeout = timeout;
+    }
+
+    public boolean isEmpty() {
+        return addresses.isEmpty();
+    }
+
+    /**
+     * Returns all addresses.
+     *
+     * @return list of all InetSocketAddress
+     */
+    public List<InetSocketAddress> getAllAddresses() {
+        return new LinkedList<>(addresses);
+    }
+
+    /**
+     * Returns wildcard addresses for all ports
+     *
+     * @return list of InetSocketAddress with wildcards for all ports
+     */
+    public List<InetSocketAddress> getWildcardAddresses() {
+       return addresses.stream().map(a -> new InetSocketAddress(a.getPort())).distinct().collect(Collectors.toList());
+    }
+
+    /**
+     * Returns all ports
+     *
+     * @return list of all ports
+     */
+    public List<Integer> getAllPorts() {
+        return addresses.stream().map(InetSocketAddress::getPort).distinct().collect(Collectors.toList());
+    }
+
+    public void addAddress(InetSocketAddress address) {
+        addresses.add(address);
+    }
+
+    /**
+     * Returns reachable address. If none is reachable than throws exception.
+     *
+     * @return address which is reachable.
+     * @throws NoRouteToHostException if none address is reachable
+     */
+    public InetSocketAddress getValidAddress() throws NoRouteToHostException {
```

Comment:
I believe `getReachableAddress` would be more suitable for this method. Every checks within refers to reachability.